### DISCUSSION TITLE
Hyperparameter Optimisation: Raytune vs no Raytune

### DISCUSSION THREAD

**Discussion by jingyu-itmax** [ultralytics/ultralytics#23729] [General]
I want to know what is the difference when we call the tune function in ultralytics, with and without raytune. Does it use the same genetic-mutation algorithm, but just runs in parallel with raytune? Or does it use Bayesian optimisation when raytune is enabled? I want to know how does ultralytics do hyperparameter optimisation, if raytune is faster why is it just not enabled defaultly.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jingyu-itmax, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly—an Ultralytics engineer will also assist soon 🙂

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (including hyperparameter tuning questions like Ray Tune vs non-Ray runs), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help us answer your Ray Tune vs non-Ray Tune question precisely, please share:
-Your `ultralytics` version (`pip show ultralytics`)
-Your OS + Python + PyTorch versions
-The exact command or Python snippet you’re using to call `tune`
-Whether you’re running single GPU, multi-GPU, or multi-node
-Any logs/output that show the tuning strategy/config being applied

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
With `model.tune(use_ray=False)` Ultralytics runs the built-in `Tuner` evolutionary search (genetic-style mutation of hypers across `iterations`, i.e. sequential “evolution” and it writes the best result to `best_hyperparameters.yaml`), while `model.tune(use_ray=True)` switches to the Ray Tune integration (`run_ray_tune`) which samples from a defined search space and uses Ray Tune’s scheduling/parallel execution (ASHA early-stopping by default) to run multiple independent training trials concurrently when you have the resources; it’s not doing Bayesian optimization unless you explicitly configure a Ray search algorithm yourself (our wrapper defaults to basic sampling + ASHA). See the [`Model.tune()` reference](https://docs.ultralytics.com/reference/engine/model/) and the [Ray Tune integration docs](https://docs.ultralytics.com/integrations/ray-tune/) for the exact behavior and parameters. Ray isn’t the default mainly because it adds a heavy optional dependency/runtime and only tends to be “faster” when you can actually parallelize trials (multi-GPU or a cluster), whereas the built-in tuner is lightweight and works out-of-the-box.

**Reply by jingyu-itmax** [reply]
Thank you for the reply. So, if raytune=False, this would be ideal for someone running experiments/projects with limited GPU resources, ```iterations``` should be low? Otherwise it will take them a long time to get the best params (unless they have the time). And with raytune, it would benefit if we set higher iterations because it is parallelisable with a large gpu cluster?

**Reply by glenn-jocher** [reply]
Yes—treat `iterations` as “number of full training trials,” so on a single/limited GPU you’ll usually want to keep `iterations` (and often `epochs`/`imgsz`) modest to stay within your time budget, then retrain once with the best settings; Ray Tune mainly helps you spend the *same* total compute more efficiently in wall-clock time via parallel trials + ASHA early-stopping, so higher `iterations` only makes sense when you actually have spare parallel capacity (and you should set `gpu_per_trial` to avoid oversubscribing). The behavior/args are documented in the [`Model.tune()` reference](https://docs.ultralytics.com/reference/engine/model/) and the [Ray Tune integration guide](https://docs.ultralytics.com/integrations/ray-tune/).
